shortfin/python/shortfin_apps/sd/components/service.py

```diff
@@ -0,0 +1,330 @@
+    ):
+        p = sf.StaticProgramParameters(self.sysman.ls, parameter_scope=parameter_scope)
+        for path in paths:
+            logging.info("Loading parameter fiber '%s' from: %s", parameter_scope, path)
+            p.load(path, format=format)
+        self.inference_parameters.append(p)
+
+    def start(self):
+        # Initiate inference programs.
+        # Resolve all function entrypoints.
+        # Start persistent processes.
+        self.batcher.launch()
+
+    def shutdown(self):
+        self.batcher.shutdown()
+
+    def __repr__(self):
+        return (
+            f"ServiceManager(\n"
+            f"  model_params={self.model_params}\n"
+            f"  inference_modules={self.inference_modules}\n"
+            f"  page_cache={self.page_cache}\n"
+            f")"
+        )
+
+
+########################################################################################
+# Batcher
+########################################################################################
+
+import math
+
+
+class BatcherProcess(sf.Process):
+    """The batcher is a persistent process responsible for flighting incoming work
+    into batches and handling the requisite cache allocations (since every batch needs
+    committed cache state).
+    """
+
+    STROBE_SHORT_DELAY = 0.1
+    STROBE_LONG_DELAY = 0.25
+
+    def __init__(self, service: GenerateService):
+        super().__init__(fiber=service.main_fiber)
+        self.service = service
+        self.batcher_infeed = self.system.create_queue()
+        self.pending_preps: set[InferenceExecRequest] = set()
+        self.pending_encodes: set[InferenceExecRequest] = set()
+        self.pending_denoises: set[InferenceExecRequest] = set()
+        self.pending_decodes: set[InferenceExecRequest] = set()
+        self.pending_postprocesses: set[InferenceExecRequest] = set()
+        self.strobe_enabled = True
+        self.strobes: int = 0
+        self.ideal_batch_size: int = max(service.model_params.submodel_batch_sizes)
+
+    def shutdown(self):
+        self.batcher_infeed.close()
+
+    def submit(self, request: StrobeMessage | InferenceExecRequest):
+        self.batcher_infeed.write_nodelay(request)
+
+    async def _background_strober(self):
+        while not self.batcher_infeed.closed:
+            await asyncio.sleep(
+                BatcherProcess.STROBE_SHORT_DELAY
+                if len(self.pending_preps) > 0
+                else BatcherProcess.STROBE_LONG_DELAY
+            )
+            if self.strobe_enabled:
+                self.submit(StrobeMessage())
+
+    async def run(self):
+        strober_task = asyncio.create_task(self._background_strober())
+        reader = self.batcher_infeed.reader()
+        while item := await reader():
+            self.strobe_enabled = False
+            if isinstance(item, InferenceExecRequest):
+                phase = item.phase
+                if phase == InferencePhase.PREP:
```

Comment:
I'm not entirely sure we need to do fine grained batching like this for this model. You could just have one exec process that said:

```
if req.needs_prep:
  await do_something()

if req.needs_encode:
  await do_something()
```

That would put some constraints on basically the batch needing to be dense across all steps but that is probably ok.  Then the batcher here is really just about lumping multiple client requests together. And if they happen to be disjoint (i.e. some need prep and others don't), then you are potentially inefficient. But I'm not sure that's really a valid case. If it is, you could just group likes together here vs enforcing the structure everywhere.

Reply:
I think this would mean we get rid of the style of resetting the request with a given inference phase, instead, we write program outputs directly to the request (not its payload attribute) and let the request keep track of its own completion with the `needs_<job>`?

Reply:
thought about this a bit more...

Regarding disjoint requests, while I don't see a client batch coming with disjoint phases as a common thing, I do imagine that a single client request may end up being disjoint in that some inference phases will be compatible for a batched inference process and others will not.
For example, we recieve a client request with 8 individual image generations which all have the same inference "parameters", i.e., they should run as batch size 8 and be coupled throughout inference execution, except they all ask for a different number of unet iterations.
This would mean that the denoising phase needs to be distributed among different program invocations, even if we do something fancy for the denoising loop and break it up into one program invocation per unet iteration. In this example, that might look like:
Do bs8 unet until one request's steps are completed
Do bs7 "
and so on...
which seems difficult unless we get really flexible about batch sizes in SD compiled modules.
Assuming we want to share work where possible, i.e., to be able to join/unjoin batches at different phases of the image generation process, is there a better way to go about it besides doing the fine-grained (phased) batching? I think we would lose quite a bit of efficiency if we had to split up batches just because they are incompatible for some subsection of the model structure -- generation parameters like inference steps, image output size, and CFG-free mode would all be examples of request parameterizations that require different programs at some point, but could be batched for other parts of the image generation process.